### PR TITLE
add yaml interpreter field

### LINKED ISSUE #89: allow to specify an interpreter for a script file
Currently `check-protocols` can only be run on executables. That means the tested files have to:

- have the executable flag set,
- have a hashbang at the top (or be elf binaries).

It should be possible to get around these restrictions by providing an interpreter, for example like this:

``` yaml
interpreter: python3
tests:
    - arguments: asdff asdfdsakl 
      protocol: 
        - asdfhhhhh
        - aljflkds

    - arguments: fjfjfjf
      protocol:
        - thing1
        - thing2

    - arguments: ofheoheff
      protocol:
        - thing1
```

The current workaround for this is to just make the scripts executable and put the hashbang in, but we don't want to force our users to run their scripts in a certain way.

cc: @xasopheno

### PR DESCRIPTION
**Pull request by data-pup**
fixes #89. :smiley_cat:

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -98,6 +98,11 @@ protocols:
         exitcode?: number
           # Mocked exitcode of the command.
           # Default: 0
+interpreter: string
+    # The interpreter that should be used to run the tested script. If no
+    # interpreter is specified, a `#!` line at the top of the script will be
+    # used to identify the interpreter.
+    # Example: "/bin/bash", default: ""
```
Comment (data-pup) [inline README.md]: here's where the new interpreter field is documented in `README`. lmk if there are any edits I should make here!

File: src/emulation/executable_mock.rs [hunk 2]
```diff
@@ -15,7 +15,7 @@ pub fn create_mock_executable(context: &Context, config: Config) -> R<Vec<u8>> {
     let mut result = b"#!".to_vec();
     result.append(
         &mut context
-            .check_protocols_executable
+            .check_protocols_program
```
Comment (data-pup) [inline src/emulation/executable_mock.rs]: because we might not be testing the behaviour of an executable if the interpreter is specified, this has been renamed.

File: src/emulation/mod.rs [hunk 3]
```diff
@@ -137,18 +139,47 @@ impl SyscallMock {
     }
 }
 
+pub enum Executable {
+    Interpreter(PathBuf),
+    Program(PathBuf),
+}
```
Comment (data-pup) [inline src/emulation/mod.rs]: This enum exists to differentiate between an executable that refers to the interpreter that should be invoked, i.e. `ruby`, and a program that is itself executable.

File: src/emulation/mod.rs [hunk 4]
```diff
@@ -137,18 +139,47 @@ impl SyscallMock {
...
+
+impl Deref for Executable {
+    type Target = Path;
+    fn deref(&self) -> &Path {
+        match self {
+            Executable::Interpreter(program) => &program,
+            Executable::Program(program) => &program,
+        }
```
Comment (data-pup) [inline src/emulation/mod.rs]: We implement `Deref` on the `Executable` enum so that we can coerce it into a `&Path` when calling `execve`

File: src/emulation/mod.rs [hunk 5]
```diff
@@ -137,18 +139,47 @@ impl SyscallMock {
...
     expected: Protocol,
     unmocked_commands: &[Vec<u8>],
 ) -> R<TestResult> {
-    let syscall_mock = Tracer::run_against_mock(
-        executable,
-        expected.arguments.clone(),
-        expected.env.clone(),
-        |tracee_pid| SyscallMock::new(context, tracee_pid, expected, unmocked_commands),
-    )?;
+    let (executable, arguments): (Executable, Vec<String>) = match interpreter {
```
Comment (data-pup) [inline src/emulation/mod.rs]: initialize the `executable` and `arguments` values depending on whether we are running a script directly, or invoking it using an interpreter

File: src/protocol/mod.rs [hunk 6]
```diff
@@ -249,6 +249,7 @@ impl Protocol {
 pub struct Protocols {
     pub protocols: Vec<Protocol>,
     pub unmocked_commands: Vec<Vec<u8>>,
+    pub interpreter: Option<Vec<u8>>,
```
Comment (data-pup) [inline src/protocol/mod.rs]: This is the new field added to `Protocols` used to represent the specified interpreter, if given

File: src/protocol/mod.rs [hunk 7]
```diff
@@ -283,11 +292,13 @@ impl Protocols {
                 (Ok(protocols), _) => {
                     let mut protocols = Protocols::from_array(protocols.expect_array()?)?;
                     protocols.add_unmocked_commands(object)?;
+                    protocols.add_interpreter(object)?;
```
Comment (data-pup) [inline src/protocol/mod.rs]: If we are given a `protocols:` object in the yaml file, here is where we add the interpreter. The function is defined above :)

File: src/protocol/mod.rs [hunk 8]
```diff
@@ -702,6 +713,51 @@ mod load {
...
+                        |interpreter: 42
+                    "##,
+                ),
+                format!(
+                    "unexpected type in {}.protocols.yaml: \
+                     expected: string, got: Integer(42)",
+                    path_to_string(&tempfile.path())?
+                )
+            );
+            Ok(())
```
Comment (data-pup) [inline src/protocol/mod.rs]: Tests to check that the interpreter parsed, and that an error is thrown if the value is not of the correct type, are defined here

File: src/syscall_mocking/mod.rs [hunk 9]
```diff
@@ -45,33 +45,42 @@ impl Tracer {
...
-            };
-            format!(
-                "execve'ing {} failed with error: {}\n{}",
-                executable.to_string_lossy(),
-                error,
-                hint,
-            )
-        })?;
+        execve(&path, &c_args, &c_env)
+            .map_err(|error| Self::format_execve_error(executable, &args, error))?;
```
Comment (data-pup) [inline src/syscall_mocking/mod.rs]: The closure for formatting an error given by `execve` has been outlined into a function above.

File: src/syscall_mocking/mod.rs [hunk 10]
```diff
@@ -45,33 +45,42 @@ impl Tracer {
...
+            }
+        };
+        format!(
+            "execve'ing {} failed with error: {}\n{}",
+            program, error, hint,
+        )
+    }
+
+    fn execve(executable: &Executable, args: Vec<String>, env: HashMap<String, String>) -> R<()> {
         let path = CString::new(executable.as_os_str().as_bytes())?;
```
Comment (data-pup) [inline src/syscall_mocking/mod.rs]: This is the line that depends on our `Deref` implementation for `Executable`.

File: src/emulation/mod.rs [hunk 11]
```diff
@@ -137,18 +139,47 @@ impl SyscallMock {
...
+        Some(interpreter) => {
+            let executable = String::from_utf8_lossy(interpreter.as_ref())
+                .into_owned()
+                .into();
+            let mut arguments = vec![path_to_string(program).unwrap().to_string()];
```
Comment (data-pup) [inline src/emulation/mod.rs]: FIXUP: use https://doc.rust-lang.org/std/path/struct.Path.html#method.to_string_lossy

thx @matthandlersux

File: src/emulation/mod.rs (hunk 4 above)
Comment (data-pup) [reply src/emulation/mod.rs]: This comment is now out of date, found a better way to do this :)

File: src/emulation/mod.rs [hunk 12]
```diff
@@ -137,18 +139,57 @@ impl SyscallMock {
...
+}
+
+impl Into<Vec<u8>> for &Executable {
+    fn into(self) -> Vec<u8> {
+        let path = match self {
+            Executable::Interpreter(path) => path,
+            Executable::Program(path) => path,
+        };
+        path.as_os_str().as_bytes().to_vec()
+    }
```
Comment (data-pup) [inline src/emulation/mod.rs]: We implement `Into<Vec<u8>>` for Executable, because the [CString constructor](https://doc.rust-lang.org/std/ffi/struct.CString.html#method.new) works on anything that implements `Into<Vec<u8>>`. I have some slightly mixed feelings about the fact that this is implemented on a reference, but changing this would mean cloning Executable afaict, which seems worse

File: src/syscall_mocking/mod.rs [hunk 13]
```diff
@@ -45,33 +47,73 @@ impl Tracer {
...
+                c_args.push_front(CString::new(interpreter.clone())?);
+                (
+                    CString::new(interpreter)?,
+                    c_args.into_iter().collect(),
+                    c_env,
+                )
+            }
+            None => (c_executable.clone(), c_args.into_iter().collect(), c_env),
+        })
+    }
```
Comment (soenkehahn) [inline src/syscall_mocking/mod.rs]: Yeah, this is actually pretty good now, I think. Pushing the interpreter name to the front of that argument list intuitively makes the most sense to me.

File: src/syscall_mocking/mod.rs [hunk 14]
```diff
@@ -45,33 +47,73 @@ impl Tracer {
...
-        })?;
+        };
+        let hint = format!("Does {} exist?", interpreter);
+        format!(
+            "execve'ing {} failed with error: {}\n{}",
+            program.to_string_lossy(),
+            error,
+            hint,
+        )
+    }
```
Comment (soenkehahn) [inline src/syscall_mocking/mod.rs]: Nice.

File: src/syscall_mocking/mod.rs [hunk 15]
```diff
@@ -45,33 +47,73 @@ impl Tracer {
...
+
+    fn execve(
+        interpreter: Option<Vec<u8>>,
```
Comment (soenkehahn) [inline src/syscall_mocking/mod.rs]: Maybe this can be passed in by reference? And then only cloned in the case of errors? Not sure whether that would make things better.

File: src/utils/mod.rs [hunk 16]
```diff
@@ -10,8 +10,9 @@ pub fn path_to_string(path: &Path) -> R<&str> {
         .ok_or_else(|| format!("invalid utf8 sequence: {:?}", &path))?)
 }
 
-pub fn parse_shebang(executable: &Path) -> Option<String> {
-    let contents = fs::read(executable).ok()?;
+pub fn parse_shebang(program: &Path) -> Option<String> {
+    let path = Path::new(program);
+    let contents = fs::read(path).ok()?;
```
Comment (soenkehahn) [inline src/utils/mod.rs]: Why was this changed?

File: tests/run.rs [hunk 17]
```diff
@@ -177,7 +194,77 @@ mod nice_user_errors {
...
+                format!(
+                    r##"
+                        |execve'ing {} failed with error: ENOEXEC: Exec format error
+                        |Does your interpreter exist?
+                    "##,
+                    path_to_string(script.path().as_ref())?
+                )
+                .as_str(),
+            )?
+            .trim()
```
Comment (soenkehahn) [inline tests/run.rs]: Why are we `trim`ming this here? Does the `result` not include a trailing `\n`? If not, shouldn't it?

Comment (soenkehahn) [reply tests/run.rs]: Errors don't have a trailing newline.
